File: services/data-api/app/text.py
"""Vietnamese-aware text normalization helpers."""
import re
import unicodedata


def strip_accents(s: str) -> str:
    s = unicodedata.normalize("NFD", str(s))
    s = "".join(c for c in s if unicodedata.category(c) != "Mn")
    return s.replace("\u0111", "d").replace("\u0110", "D")


def norm_key(s: str) -> str:
    """Aggressive key for matching photo folder names (ignore punctuation)."""
    return re.sub(r"[^a-z0-9]+", "", strip_accents(s).lower())


def norm_cmp(s) -> str:
    """Looser key for comparing filter values (keep word boundaries)."""
    if s is None:
        return ""
    return " ".join(strip_accents(s).lower().split())

File: apps/desktop/src/lib/text.ts
/** Vietnamese-aware text normalization helpers. */

export function stripDiacritics(s: string): string {
  return s
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/đ/g, "d")
    .replace(/Đ/g, "D");
}

/** Lowercase, drop diacritics and non-alphanumerics (for fuzzy matching). */
export function norm(s: unknown): string {
  return stripDiacritics(String(s ?? "")).toLowerCase().replace(/[^a-z0-9]+/g, "");
}

export function normCompare(a: unknown, b: unknown): boolean {
  return norm(a) === norm(b);
}
